src/workerd/io/io-context.h

```diff
@@ -901,200 +892,166 @@ kj::PromiseForResult<Func, Worker::Lock&> IoContext::run(
+          // a JS Error here, it won't have a useful stack trace. V8 can generate a good stack
+          // trace as long as we construct the Error inside of a promise continuation, so we use
+          // a `.then()` below that acutally extracts the kj::Exception and turn it into a JS
+          // Error.
+          resolver.resolve(kj::mv(e));
+        } else {
+          try {
+            js.tryCatch([&]() {
+              if constexpr (jsg::isVoid<Result>()) {
+                func(js);
+                resolver.resolve(js, kj::none);
+              } else {
+                resolver.resolve(js, func(js));
+              }
+            }, [&](jsg::Value error) {
+              // Here we can just `resolver.reject` because we already have a JS exception.
+              resolver.reject(js, error.getHandle(js));
+            });
+          } catch (...) {
+            // Again, pass along the KJ exception so we can convert it later in the right context.
+            resolver.resolve(js, kj::getCaughtExceptionAsKj());
           }
+        }
+      } else {
+        // T is not void.
+        KJ_SWITCH_ONEOF(exceptionOrT) {
           KJ_CASE_ONEOF(exception, kj::Exception) {
-            // We don't use `resolver.reject()` here because if we convert the kj::Exception into
-            // a JS Error here, it won't have a useful stack trace. V8 can generate a good stack
-            // trace as long as we construct the Error inside of a promise continuation, so we use
-            // a `.then()` below that acutally extracts the kj::Exception and turn it into a JS
-            // Error.
-            resolver.resolve(kj::mv(exception));
+            // Again, pass along the KJ exception so we can convert it later in the right context.
+            resolver.resolve(js, kj::mv(exception));
+          }
+          KJ_CASE_ONEOF(result, T) {
+            try {
+              js.tryCatch([&]() {
+                if constexpr (jsg::isVoid<Result>()) {
+                  func(js, kj::mv(result));
+                  resolver.resolve(js, kj::none);
+                } else {
+                  // Here we can just `resolver.reject` because we already have a JS exception.
+                  resolver.resolve(js, func(js, kj::mv(result)));
+                }
+              }, [&](jsg::Value error) {
+                resolver.reject(js, error.getHandle(js));
+              });
+            } catch (...) {
+              // Again, pass along the KJ exception so we can convert it later in the right context.
+              resolver.resolve(js, kj::getCaughtExceptionAsKj());
+            }
           }
         }
-      }, kj::mv(ilOrCs));
-    }));
+      }
+    }, kj::mv(ilOrCs));
+  }));
 
-    return jsPromise.then(js, [](jsg::Lock& js, IoResultOrException&& resultOrException) -> Result {
-      KJ_SWITCH_ONEOF(resultOrException) {
+  return jsPromise.then(js, [](jsg::Lock& js, ExceptionOr<Result>&& exceptionOrResult) -> Result {
```

Comment:
non-blocking: it might be worthwhile adding a short comment here reminding folks that the `then` call here could throw a `JsExceptionThrown` but that it's ok if it does since `func` is already attached. Completely optional tho ... I just want to see about taking opportunities to remind folks that the `jsg::Promise` `then`/`catch_` calls can throw and need to handled appropriately.

Reply:
07f8cb10